src/features/caip/tokens.ts

```diff
@@ -26,27 +26,38 @@ export function getCaip19Id(
   if (!isValidAddress(address) && !isZeroishAddress(address)) {
     throw new Error(`Invalid address: ${address}`);
   }
-  return `${chainCaip2Id}/${namespace}:${address}${tokenId ? `/${tokenId}` : ''}`;
+  // NOTE: deviation from CAIP-19 spec here by separating token id with : instead of /
+  // Doing this because cosmos addresses use / all over the place
+  // The CAIP standard doesn't specify how to handle ibc / token factory addresses
+  return `${chainCaip2Id}/${namespace}:${address}${tokenId ? `:${tokenId}` : ''}`;
 }
 
 export function parseCaip19Id(id: TokenCaip19Id) {
   const segments = id.split('/');
-  if (segments.length >= 2) {
-    const chainCaip2Id = segments[0] as ChainCaip2Id;
-    const isIBCDenom = segments[1] === `${AssetNamespace.ibcDenom}:ibc`;
-
-    const [namespace, address] = isIBCDenom
-      ? [AssetNamespace.ibcDenom, `ibc/${segments[2]}`]
-      : (segments[1].split(':') as [AssetNamespace, Address]);
-
-    if (!chainCaip2Id || !namespace || !address) {
-      throw new Error(`Invalid caip19 id: ${id}`);
-    }
-    const tokenId = segments.length > 2 ? segments[2] : undefined;
-    return { chainCaip2Id, namespace, address, tokenId };
+  if (segments.length < 2)
+    throw new Error(`Invalid caip19 id: ${id}. Must have at least 2 main segments`);
+
+  const chainCaip2Id = segments[0] as ChainCaip2Id;
+  const rest = segments.slice(1).join('/');
+  const tokenSegments = rest.split(':');
+  let namespace: AssetNamespace;
+  let address: Address;
+  let tokenId: string | undefined;
+  if (tokenSegments.length == 2) {
+    [namespace, address] = tokenSegments as [AssetNamespace, Address];
+  } else if (tokenSegments.length == 3) {
+    // NOTE: deviation from CAIP-19 spec here by separating token id with : instead of /
+    // Doing this because cosmos addresses use / all over the place
+    // The CAIP standard doesn't specify how to handle ibc / token factory addresses
+    [namespace, address, tokenId] = tokenSegments as [AssetNamespace, Address, string];
```

Comment:
do we have a zod parser for these instead of casts?

Reply:
I don't think Zod would be a good fit for this kind of nuanced splitting and parsing but I'm open to revisiting this code if you disagree